### ISSUE TITLE
Signatures should appear above excerpts

### ISSUE THREAD

**Issue by Radvendii** [astroidmail/astroid#456]
If I reply or forward a message, I usually want the signature to appear above the quoted message. I assume this is the usual desire, but now that I write this I'm not sure. Do people usually want their signature at the end of the whole email, or at the end of their part of the email?

The thing is, sometimes I intersperse my writing with quoted material, and in that case I want my signature to go at the end.

I'm guessing the right solution to this is to check if the email has a block of quoted text at the very end of it, and if so put the signature before it.

**Comment by emdete** [comment]
this seems to be a request for tofu (Text Over, Fullquote Under) which is one of the religious wars in emailing. so once considering it should be at least configurable.

**Comment by gauteh** [comment]
Another issue with this is that the message body and the signature are 
not easily separable. The current behavior is easy to implement since it 
just means add the signature at the end - and keep the body separate. 
The suggested behavior would require the signature to be added inline, 
essentially making it very hard to toggle it from astroid! (still easy 
enough to just delete the lines in the editor..)

**Comment by Radvendii** [comment]
@emdete Oh okay I didn't realize this was a debated issue, though it makes sense that it is
@gauteh Ohhh good call. I didn't realize astroid just stuck the lines in the file. Still, it should be easy enough to just do the same thing, but up past the last block quote, no?

**Comment by gauteh** [comment]
Taeer Bar-Yam writes on januar 24, 2018 14:41:
> @gauteh Ohhh good call. I didn't realize astroid just stuck the lines in the file. Still, it should be easy enough to just do the same thing, but up past the last block quote, no?

True, should be possible - but would require a good definition of when 
quote starts + parsing, and also define what happens if no quote is 
found. 

What should happen if you have several inline quotes? Or the message 
starts with a quote?

**Comment by hugoroy** [comment]
OK, maybe not a good idea because it is radically different from what we do for signatures currently, but:

What about making a "tofu" option? Astroid would just need to rely on where the quote/forward line (that are already configurable) is to know where to put the signature, and not have to parse the quotes.

**Comment by gauteh** [comment]
I think the easiest way to implement this in a clean way is to:

* Configurably insert signature before quote line (essentially at the 
  top) when a new draft is first created. Should there be any rules for 
  when this should be done as opposed to leaving it at the bottom?

* If insert at top, the signature will _not_ be toggleable. You will 
  have to remove the signature manually in the editor.

* If configured off; behavior is as it is now.

**Comment by hugoroy** [comment]
Yes I think this makes sense to do it like this if we want to enable that feature.


Le 6 février 2018 18:59:48 GMT+01:00, Gaute Hope <notifications@github.com> a écrit :
>I think the easiest way to implement this in a clean way is to:
>
>* Configurably insert signature before quote line (essentially at the 
> top) when a new draft is first created. Should there be any rules for 
>  when this should be done as opposed to leaving it at the bottom?
>
>* If insert at top, the signature will _not_ be toggleable. You will 
>  have to remove the signature manually in the editor.
>
>* If configured off; behavior is as it is now.

-- 
Envoyé de mon appareil mobile
Sent from my mobile device
